targets/TARGET_Realtek/TARGET_AMEBA/rtw_emac.cpp

```diff
@@ -73,80 +78,93 @@ static void wlan_get_hwaddr(emac_interface_t *emac, uint8_t *addr)
     }
 }
 
-static void wlan_set_hwaddr(emac_interface_t *emac, uint8_t *addr)
+void RTW_EMAC::set_hwaddr(const uint8_t *addr)
 {
-    
 }
 
-
-static bool wlan_link_out(emac_interface_t *emac, emac_stack_mem_t *buf)
+bool RTW_EMAC::link_out(emac_mem_buf_t *buf)
 {
-    struct eth_drv_sg * sg_list=0;
+    struct eth_drv_sg *sg_list;
     int sg_len = 0;
     int tot_len;
-    struct pbuf *p;
+    emac_mem_buf_t *p;
     bool ret = true;
-
     if (!rltk_wlan_running(0)) {
+        memory_manager->free(buf);
         return false;
     }
 
     sg_list = (struct eth_drv_sg *)malloc(sizeof(struct eth_drv_sg)*MAX_ETH_DRV_SG);
-    if(sg_list == 0){//malloc fail
+    if (sg_list == 0) {
+        memory_manager->free(buf);
         return false;
     }
-    emac_stack_mem_ref(emac, buf);
-
-    p = (struct pbuf *)buf;
-    tot_len = p->tot_len;
-    for (; p != NULL && sg_len < MAX_ETH_DRV_SG; p = p->next) {
-        sg_list[sg_len].buf = (uint32_t) p->payload;
-        sg_list[sg_len].len = p->len;
-    sg_len++;
-    }
 
+    p = buf;
+    tot_len = memory_manager->get_total_len(p);
+    for (; p != NULL && sg_len < MAX_ETH_DRV_SG; p = memory_manager->get_next(p)) {
+        sg_list[sg_len].buf = (unsigned int)(static_cast<uint8_t *>(memory_manager->get_ptr(p)));
+        sg_list[sg_len].len = memory_manager->get_len(p);
+        sg_len++;
+    }
     if (sg_len) {
         if (rltk_wlan_send(0, sg_list, sg_len, tot_len) != 0) {
             ret = false;
         }
     }
 
-    emac_stack_mem_free(emac, buf);
+    memory_manager->free(buf);
     free(sg_list);
     return ret;
 }
 
-static bool wlan_power_up(emac_interface_t *emac)
+bool RTW_EMAC::power_up()
 {
     wifi_on(RTW_MODE_STA);
     wait_ms(1000);
```

Comment:
Link up can be made here. e.g. 
    if (emac_link_state_cb) {
        emac_link_state_cb(true);
    }

Reply:
@mikaleppanen 
Do yo mean wlan_emac_link_change(true); ?
emac_link_state_cb is another callback.

Reply:
emac_link_state_cb is the callback set by the stack, so it can be called here.

Reply:
Confirmed to use wlan_emac_link_change(true); 
wlan_emac_link_change(true); and 
if (emac_link_state_cb) {
emac_link_state_cb(true);
} 
actually are same. Thank you.